analyzers/src/SonarAnalyzer.CSharp/Rules/ClassNotInstantiatable.cs

```diff
@@ -45,6 +43,9 @@ public sealed class ClassNotInstantiatable : ClassNotInstantiatableBase
         protected override bool IsTypeDeclaration(SyntaxNode node) =>
             node.IsAnyKind(SyntaxKind.ClassDeclaration, SyntaxKindEx.RecordDeclaration);
 
+        protected override bool IsObjectCreation(SyntaxNode node) =>
+            node.IsAnyKind(SyntaxKind.ObjectCreationExpression, SyntaxKindEx.ImplicitObjectCreationExpression);
+
         private void CheckClassWithOnlyUnusedPrivateConstructors(SymbolAnalysisContext context)
```

Comment:
Let's refactor this rule to remove duplications. 
`Initialize` can be moved to base class.
`Rule` and `SupportedDiagnostics` can be moved to base class (rule instance will be created in the constructor)
`CheckClassWithOnlyUnusedPrivateConstructors` can be moved to base class.

There will be just few minor things to deal with as overridables - `CollectRemovableDeclarations` - it can return pairs of `name` and `typeDeclaration` for the error message.

Reply:
See `DoNotLockWeakIdentityObjectsBase` (for example) for inspiration/snippets